staging/src/k8s.io/client-go/features/env_var_feature_gate.go

```diff
@@ -0,0 +1,121 @@
+
+// newEnvVarFeatureGate returns a new Reader that recognizes the provided features.
+//
+// On the first call to Enabled, the effective state of all known features is loaded from
+// environment variables. The environment variable read for a given feature is formed by
+// concatenating the prefix "KUBE_FEATURE_" with the feature's name.
+//
+// For example, if you have a feature named "MyFeature,"
+// setting an environmental variable "KUBE_FEATURE_MyFeature"
+// will allow you to configure the state of that feature.
+//
+// Please note that environmental variables can only be set to a boolean type.
+// Incorrect values will be ignored and logged.
+func newEnvVarFeatureGate(features map[Feature]FeatureSpec) *envVarFeatureGate {
+	enabled := map[Feature]bool{}
+	enabledValue := &atomic.Value{}
+	enabledValue.Store(enabled)
+
+	known := map[Feature]FeatureSpec{}
+	for name, spec := range features {
+		known[name] = spec
+	}
+
+	return &envVarFeatureGate{
+		envVarFeatureGateName: naming.GetNameFromCallsite(internalPackages...),
+		known:                 known,
+		enabled:               enabledValue,
+	}
+}
+
+// envVarFeatureGate implements Reader and allows for feature registration.
+type envVarFeatureGate struct {
+	// envVarFeatureGateName holds the name of the file
+	// that created this instance
+	envVarFeatureGateName string
+
+	// readEnvVarsOnce guards reading environmental variables
+	readEnvVarsOnce sync.Once
+
+	// known holds known feature gates
+	known map[Feature]FeatureSpec
+
+	// enabled holds a map[Feature]bool
+	// with values explicitly set via env var
+	enabled *atomic.Value
+}
+
+// Enabled returns true if the key is enabled.  If the key is not known, this call will panic.
+func (f *envVarFeatureGate) Enabled(key Feature) bool {
+	if v, ok := f.getEnabledMapFromEnvVar()[key]; ok {
+		return v
+	}
+	if v, ok := f.known[key]; ok {
+		return v.Default
+	}
+	panic(fmt.Errorf("feature %q is not registered in FeatureGate %q", key, f.envVarFeatureGateName))
+}
+
+// getEnabledMapFromEnvVar will fill the enabled map on the first call.
+// This is the only time a known feature can be set to a value
+// read from the corresponding environmental variable.
+func (f *envVarFeatureGate) getEnabledMapFromEnvVar() map[Feature]bool {
+	f.readEnvVarsOnce.Do(func() {
+		featureGateState := map[Feature]bool{}
+		for feature, featureSpec := range f.known {
+			featureState, featureStateSet := os.LookupEnv(fmt.Sprintf("KUBE_FEATURE_%s", feature))
+			if !featureStateSet {
+				continue
+			}
+			boolVal, boolErr := strconv.ParseBool(featureState)
+			switch {
+			case boolErr != nil:
+				utilruntime.HandleError(fmt.Errorf("cannot set feature gate %v to %v, due to %v", feature, featureState, boolErr))
+			case featureSpec.LockToDefault:
+				if boolVal != featureSpec.Default {
+					utilruntime.HandleError(fmt.Errorf("cannot set feature gate %v to %v, feature is locked to %v", feature, featureState, featureSpec.Default))
+					break
+				}
```

Comment:
The main purpose of this function is to give admins a way to disable faulty behaviour. The primary users are kube components which will expose client-go's gates as command line options (potential parsing errors will be reported).  Possibly in a phased manner, initially a feature will be enabled only for a limited number of components. Hopefully this will give developers enough time to test their features  and remove the gates altogether.

For components that wish to test a feature early and do not expose flags as command line options, any potential parsing errors will be logged into a log file, and the default value will be applied. We could change the way we report parsing errors by:
-  crashing the entire program on errors but this approach seems to be way too powerful.
-  exposing a new method that would return any errors and rely on binaries to make use of the new function.

Note that the current behaviour is similar when users set `KUBE_CLIENT_BACKOFF_BASE` and `KUBE_CLIENT_BACKOFF_DURATION`
https://github.com/kubernetes/kubernetes/blob/78cb3862f11225135afdf76f3424e2d7b33104c7/staging/src/k8s.io/client-go/rest/client.go#L148

Reply:
I'm hesitant to suggest a panic too, but it has to be weighed against the scenario where a severe bug is discovered in a default-enabled client feature (say for example a bug that can result in data loss/corruption). If an administrator tries to disable the feature by setting the environment variable `KUBE_FEATURE_Foo=fales` instead of the correct `KUBE_FEATURE_Foo=false`, we can be reasonably sure that their intention was to set an override for the feature named Foo, and they are likely to believe that they have successfully mitigated the bug when in fact nothing has changed and the impact of the bug is prolonged.

Sorry for equivocating. I think the additional logging you suggested is a good idea and provides a way for users to verify that feature gate states are as they expect in non-kube components.

Reply:
Since client-go is a library, our options are limited, assuming we would like to auto-detect parsing issues.

Another option that came to my mind would be to slightly extend the `Reader` interface to:
```
type Reader interface {
    Enabled(string) (bool, error)
}
```

The additional error could be used to report potential issues, but this solution wouldn't report parsing issues in a clear way, so I don't like this approach.

I also think that checking the additional logging for 3rd party components is a reasonable ask.

Reply:
> I also think that checking the additional logging for 3rd party components is a reasonable ask.

Agreed. I think the behavior you have here is reasonable given the constraints that come with being a library, especially if it's trivial for administrators to check the effective feature state.